File: Python/navio/gpio.py
import glob

import gpiod
from gpiod.line import Direction, Value


def _find_line(pin):
    """Locate the header GPIO line named GPIO<pin>.

    On Raspberry Pi 5 the header GPIOs are on the RP1 gpiochip, on older
    boards on the BCM pinctrl chip - on both, lines are named "GPIO<n>".
    """
    name = "GPIO%d" % pin
    for path in sorted(glob.glob("/dev/gpiochip*")):
        try:
            with gpiod.Chip(path) as chip:
                offset = chip.line_offset_from_id(name)
                return path, offset
        except (ValueError, OSError, PermissionError):
            continue
    raise RuntimeError("GPIO line %s not found (are you root?)" % name)


class Pin():
    """GPIO pin wrapper backed by the gpiod character device."""

    def __init__(self, pin):
        self.pin = pin
        self._path, self._offset = _find_line(pin)
        self._request = gpiod.request_lines(
            self._path,
            consumer="navio",
            config={self._offset: gpiod.LineSettings(direction=Direction.INPUT)},
        )
        self._direction = Direction.INPUT

    def _set_direction(self, direction):
        if self._direction != direction:
            self._request.reconfigure_lines(
                {self._offset: gpiod.LineSettings(direction=direction)}
            )
            self._direction = direction

    def write(self, value):
        self._set_direction(Direction.OUTPUT)
        self._request.set_value(
            self._offset, Value.ACTIVE if int(value) else Value.INACTIVE
        )

    def read(self):
        return 1 if self._request.get_value(self._offset) == Value.ACTIVE else 0


if __name__ == "__main__":
    pin = Pin(27)
    pin.write(0)

File: C++/Navio/gpio.cpp
#include <cstdio>
#include <err.h>
#include <gpiod.h>

#include "gpio.h"

#define LOW  0
#define HIGH 1

using namespace Navio;

Pin::Pin(uint8_t pin):
    _pin(pin),
    _mode(GpioModeInput),
    _chip(NULL),
    _line(NULL)
{
}

Pin::~Pin()
{
    if (_line) {
        gpiod_line_release(_line);
    }
    if (_chip) {
        gpiod_chip_close(_chip);
    }
}

bool Pin::init()
{
    char line_name[16];
    snprintf(line_name, sizeof(line_name), "GPIO%u", _pin);

    /* Search every gpiochip for the header line with this name.
       On Pi 5 the header GPIOs are provided by the RP1 (label pinctrl-rp1),
       on older boards by the BCM pinctrl - line names match on both. */
    struct gpiod_chip_iter *iter = gpiod_chip_iter_new();
    if (!iter) {
        warnx("cannot iterate gpio chips (are you root?)");
        return false;
    }

    struct gpiod_chip *chip;
    gpiod_foreach_chip(iter, chip) {
        struct gpiod_line *line = gpiod_chip_find_line(chip, line_name);
        if (line) {
            _chip = chip;
            _line = line;
            break;
        }
    }

    if (_line) {
        gpiod_chip_iter_free_noclose(iter);
    } else {
        gpiod_chip_iter_free(iter);
        warnx("gpio line %s not found", line_name);
        return false;
    }

    return _request(_mode);
}

bool Pin::_request(GpioMode mode)
{
    gpiod_line_release(_line);

    int ret;
    if (mode == GpioModeOutput) {
        ret = gpiod_line_request_output(_line, "navio", 0);
    } else {
        ret = gpiod_line_request_input(_line, "navio");
    }

    if (ret < 0) {
        warn("cannot request gpio line %u", _pin);
        return false;
    }

    return true;
}

void Pin::setMode(GpioMode mode)
{
    if (_line && _request(mode)) {
        _mode = mode;
    }
}

uint8_t Pin::read() const
{
    if (!_line) {
        return 0;
    }
    int value = gpiod_line_get_value(_line);
    return value > 0 ? 1 : 0;
}

void Pin::write(uint8_t value)
{
    if (_mode != GpioModeOutput) {
        warnx("no effect because mode is not set");
        return;
    }
    if (_line) {
        gpiod_line_set_value(_line, value == LOW ? 0 : 1);
    }
}

void Pin::toggle()
{
    write(!read());
}
